### ISSUE TITLE
Installing TorchRec results in '_OpNamespace' object has no attribute 'jagged_2d_to_dense'

### ISSUE THREAD

**Issue by janekl** [meta-pytorch/torchrec#839]
Hello. I'm trying to install TorchRec inside `nvcr.io/nvidia/pytorch:22.07-py3` container that comes with CUDA 11.7. The installation itself looks successful but when I try to do `import torchrec` in Python later I get some errors that apparently are related to `fbgemm_gpu`  package.

The simplest reproducibility instruction I can offer is:
```
docker run nvcr.io/nvidia/pytorch:22.07-py3 bash -c 'pip install torchrec && python -c "import torchrec"'
```

The error message is:
```
libtorch_cuda_cpp.so: cannot open shared object file: No such file or directory
Traceback (most recent call last):
  File "/opt/conda/lib/python3.8/site-packages/torch/_ops.py", line 203, in __getattr__
    op, overload_names = torch._C._jit_get_operation(qualified_op_name)
RuntimeError: No such operator fbgemm::jagged_2d_to_dense

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/opt/conda/lib/python3.8/site-packages/torchrec/__init__.py", line 8, in <module>
    import torchrec.distributed  # noqa
  File "/opt/conda/lib/python3.8/site-packages/torchrec/distributed/__init__.py", line 36, in <module>
    from torchrec.distributed.model_parallel import DistributedModelParallel  # noqa
  File "/opt/conda/lib/python3.8/site-packages/torchrec/distributed/model_parallel.py", line 21, in <module>
    from torchrec.distributed.planner import (
  File "/opt/conda/lib/python3.8/site-packages/torchrec/distributed/planner/__init__.py", line 22, in <module>
    from torchrec.distributed.planner.planners import EmbeddingShardingPlanner  # noqa
  File "/opt/conda/lib/python3.8/site-packages/torchrec/distributed/planner/planners.py", line 16, in <module>
    from torchrec.distributed.planner.constants import BATCH_SIZE, MAX_SIZE
  File "/opt/conda/lib/python3.8/site-packages/torchrec/distributed/planner/constants.py", line 10, in <module>
    from torchrec.distributed.embedding_types import EmbeddingComputeKernel
  File "/opt/conda/lib/python3.8/site-packages/torchrec/distributed/embedding_types.py", line 14, in <module>
    from fbgemm_gpu.split_table_batched_embeddings_ops import EmbeddingLocation
  File "/opt/conda/lib/python3.8/site-packages/fbgemm_gpu/__init__.py", line 22, in <module>
    from . import _fbgemm_gpu_docs
  File "/opt/conda/lib/python3.8/site-packages/fbgemm_gpu/_fbgemm_gpu_docs.py", line 18, in <module>
    torch.ops.fbgemm.jagged_2d_to_dense,
  File "/opt/conda/lib/python3.8/site-packages/torch/_ops.py", line 207, in __getattr__
    raise AttributeError(f"'_OpNamespace' object has no attribute '{op_name}'") from e
AttributeError: '_OpNamespace' object has no attribute 'jagged_2d_to_dense'
```

Some details on library versions (from `pip freeze`):
```
torch==1.13.0a0+08820cb
torchrec==0.3.1 
fbgemm-gpu==0.3.0
```

Do you have any idea what goes wrong here?

**Comment by YLGH** [comment]
please install the nightly version of fbgemm-gpu if you're using torch-nightly

I think

```
pip uninstall fbgemm-gpu
pip install fbgemm-gpu-nightly
```

should work

**Comment by osalpekar** [comment]
Echoing @YLGH's comment - that error is seen upon importing torchrec when torchrec is built with fbgemm-gpu's stable release and torch-nightly. One alternative may be to set `CHANNEL=nightly` when building torchrec, and torchrec's `setup.py` should automatically install fbgemm-gpu-nightly during the build process.

**Comment by janekl** [comment]
Thank you for your quick response!

So now I tried both
1. installing `fbgemm-gpu`, uninstalling `fbgemm-gpu` and then installing `fbgemm-gpu-nightly`
2. installing` fbgemm-gpu-nightly` (that installs `fbgemm-gpu-nightly` itself)

But I'm still hitting some issues in either case. For Option (1) I have the following library versions (I'm using `nvcr.io/nvidia/pytorch:22.07-py3` container as above):
``` 
torch==1.13.0a0+08820cb
torchrec==0.3.1
fbgemm-gpu-nightly==2022.11.28
```

Now, when I try to import `torchrec` I get the following:
```
$ python
Python 3.8.13 | packaged by conda-forge | (default, Mar 25 2022, 06:04:10)
[GCC 10.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import torchrec
/opt/conda/lib/python3.8/site-packages/fbgemm_gpu/fbgemm_gpu_py.so: undefined symbol: _ZNK5torch8autograd4Node4nameEv
>>> torchrec
<module 'torchrec' from '/opt/conda/lib/python3.8/site-packages/torchrec/__init__.py'>
```

Apparently the import is successful, but there is "undefined symbol" message which might be a symptom that something may still be incorrect. And when I'm trying to use the library with [torchrec_dlrm](https://github.com/facebookresearch/dlrm/tree/main/torchrec_dlrm) example then I'm facing the following issues:

1. For unit tests with `python -m unittest torchrec_dlrm/tests/test_dlrm_main.py`:
```
Traceback (most recent call last):
  File "/opt/conda/lib/python3.8/site-packages/torch/_ops.py", line 203, in __getattr__
    op, overload_names = torch._C._jit_get_operation(qualified_op_name)
RuntimeError: No such operator fbgemm::asynchronous_complete_cumsum

...

Traceback (most recent call last):
  File "/opt/conda/lib/python3.8/site-packages/torch/_ops.py", line 203, in __getattr__
    op, overload_names = torch._C._jit_get_operation(qualified_op_name)
RuntimeError: No such operator fbgemm::dense_embedding_codegen_lookup_function
```

2. For a more complex example when running [dlrm_main.py script](https://github.com/facebookresearch/dlrm/blob/main/torchrec_dlrm/dlrm_main.py) (for 8xA100 GPUs setup):
```
torchx run -s local_cwd dist.ddp -j 1x8 --script dlrm_main.py -- \
    --embedding_dim 128 \
    --dense_arch_layer_sizes 512,256,128 \
    --over_arch_layer_sizes 1024,1024,512,256,1 \
    --in_memory_binary_criteo_path /data \
    --epochs 1 \
    --pin_memory \
    --mmap_mode \
    --batch_size 2048 \
    --test_batch_size 131072 \
    --interaction_type dcn \
    --dcn_num_layers 3 \
    --dcn_low_rank_dim 512 \
    --adagrad \
    --learning_rate 0.005 \
    --limit_test_batches 100 \
    --limit_train_batches 4000 \
    --limit_val_batches 20
```
I'm hitting new issues- this time it says:
```
dlrm_main/0 [0]:Traceback (most recent call last):
dlrm_main/0 [0]:  File "/opt/conda/lib/python3.8/site-packages/torch/_ops.py", line 203, in __getattr__
dlrm_main/0 [0]:    op, overload_names = torch._C._jit_get_operation(qualified_op_name)
dlrm_main/0 [0]:RuntimeError: No such operator fbgemm::permute_2D_sparse_data
```

Everything looks related to `fbgemm_gpu`  installation. I'm observing the same issues when following setup Option (2) above (that is, when both `torchrec` and `fbgemm_gpu` are nightly versions).

Would you have any other suggestions how to fix this?

**Comment by YLGH** [comment]
> Thank you for your quick response!
> 
> So now I tried both
> 
> 1. installing `fbgemm-gpu`, uninstalling `fbgemm-gpu` and then installing `fbgemm-gpu-nightly`
> 2. installing` fbgemm-gpu-nightly` (that installs `fbgemm-gpu-nightly` itself)
> 
> But I'm still hitting some issues in either case. For Option (1) I have the following library versions (I'm using `nvcr.io/nvidia/pytorch:22.07-py3` container as above):
> 
> ```
> torch==1.13.0a0+08820cb
> torchrec==0.3.1
> fbgemm-gpu-nightly==2022.11.28
> ```
> 
> Now, when I try to import `torchrec` I get the following:
> 
> ```
> $ python
> Python 3.8.13 | packaged by conda-forge | (default, Mar 25 2022, 06:04:10)
> [GCC 10.3.0] on linux
> Type "help", "copyright", "credits" or "license" for more information.
> >>> import torchrec
> /opt/conda/lib/python3.8/site-packages/fbgemm_gpu/fbgemm_gpu_py.so: undefined symbol: _ZNK5torch8autograd4Node4nameEv
> >>> torchrec
> <module 'torchrec' from '/opt/conda/lib/python3.8/site-packages/torchrec/__init__.py'>
> ```
> 
> Apparently the import is successful, but there is "undefined symbol" message which might be a symptom that something may still be incorrect. And when I'm trying to use the library with [torchrec_dlrm](https://github.com/facebookresearch/dlrm/tree/main/torchrec_dlrm) example then I'm facing the following issues:
> 
> 1. For unit tests with `python -m unittest torchrec_dlrm/tests/test_dlrm_main.py`:
> 
> ```
> Traceback (most recent call last):
>   File "/opt/conda/lib/python3.8/site-packages/torch/_ops.py", line 203, in __getattr__
>     op, overload_names = torch._C._jit_get_operation(qualified_op_name)
> RuntimeError: No such operator fbgemm::asynchronous_complete_cumsum
> 
> ...
> 
> Traceback (most recent call last):
>   File "/opt/conda/lib/python3.8/site-packages/torch/_ops.py", line 203, in __getattr__
>     op, overload_names = torch._C._jit_get_operation(qualified_op_name)
> RuntimeError: No such operator fbgemm::dense_embedding_codegen_lookup_function
> ```
> 
> 2. For a more complex example when running [dlrm_main.py script](https://github.com/facebookresearch/dlrm/blob/main/torchrec_dlrm/dlrm_main.py) (for 8xA100 GPUs setup):
> 
> ```
> torchx run -s local_cwd dist.ddp -j 1x8 --script dlrm_main.py -- \
>     --embedding_dim 128 \
>     --dense_arch_layer_sizes 512,256,128 \
>     --over_arch_layer_sizes 1024,1024,512,256,1 \
>     --in_memory_binary_criteo_path /data \
>     --epochs 1 \
>     --pin_memory \
>     --mmap_mode \
>     --batch_size 2048 \
>     --test_batch_size 131072 \
>     --interaction_type dcn \
>     --dcn_num_layers 3 \
>     --dcn_low_rank_dim 512 \
>     --adagrad \
>     --learning_rate 0.005 \
>     --limit_test_batches 100 \
>     --limit_train_batches 4000 \
>     --limit_val_batches 20
> ```
> 
> I'm hitting new issues- this time it says:
> 
> ```
> dlrm_main/0 [0]:Traceback (most recent call last):
> dlrm_main/0 [0]:  File "/opt/conda/lib/python3.8/site-packages/torch/_ops.py", line 203, in __getattr__
> dlrm_main/0 [0]:    op, overload_names = torch._C._jit_get_operation(qualified_op_name)
> dlrm_main/0 [0]:RuntimeError: No such operator fbgemm::permute_2D_sparse_data
> ```
> 
> Everything looks related to `fbgemm_gpu` installation. I'm observing the same issues when following setup Option (2) above (that is, when both `torchrec` and `fbgemm_gpu` are nightly versions).
> 
> Would you have any other suggestions how to fix this?

I think the latest FBGEMM nightly versions could have some issues. Could you try installing with an older version of FBGEMM? 

```
pip uninstall fbgemm_gpu/fbgemm_gpu_nightly
pip install fbgemm-gpu-nightly==2022.11.22
```

or around that time?

**Comment by samiwilf** [comment]
@janekl I was able to reproduce the issue and resolve it.  
Run:
pip uninstall torch
pip3 install --pre torch torchvision torchaudio --extra-index-url https://download.pytorch.org/whl/nightly/cu117

That should resolve the issue.  All three (torch, fbgemm_gpu, and torchrec) should be nightly versions.

One sidenote, latest changes adding support for shuffle on criteo load was added on 11.24, which is after latest torchrec-nightly release dated 2022.11.21.  Checking out the prior commit in the facebookresearch/dlrm repo will resolve that.

**Comment by janekl** [comment]
I checked today with the [versions available](https://pypi.org/project/fbgemm-gpu-nightly/#history) on pip today and with both fbgemm-gpu-nightly==2022.11.30 and 2022.12.7 I'm still getting the same error on `import fbgemm_gpu`:
```
/opt/conda/lib/python3.8/site-packages/fbgemm_gpu/fbgemm_gpu_py.so: undefined symbol: _ZNK5torch8autograd4Node4nameEv
```

The container is [PyTorch Release 22.07](https://docs.nvidia.com/deeplearning/frameworks/pytorch-release-notes/rel_22-07.html#rel_22-07).

I also checked with the latest container available `nvcr.io/nvidia/pytorch:22.11-py3` and the problem persists.

I wonder that maybe this is more of `fbgemm_gpu` than a `torchrec` issue so I'm not sure now it's the right place to discuss this..

Besides, I think that what you @samiwilf suggested in your last post _should_ work: but my point is how to make it work with the Pytorch version included in the NGC containers to avoid reinstalling it.
